src/main/java/io/vertx/core/eventbus/impl/Handlers.java

```diff
@@ -34,12 +70,21 @@ public HandlerHolder choose() {
         pos.set(0);
       }
       try {
-        return list.get(p);
-      } catch (IndexOutOfBoundsException e) {
+        return elements[p];
+      } catch (ArrayIndexOutOfBoundsException e) {
```

Comment:
we need to reason about it, I would say it might because of a concurrent update with the Atomic position but I haven't checked

Reply:
I think now it's still needed because two threads might race on 

```
      int p = pos.getAndIncrement();
      if (p >= size - 1) {
        pos.set(0);
      }
```

and have `pos >= size`

to avoid it we would need to do a CAS certainly

Reply:
:+1: for CAS

Reply:
CAS is not free :-)

Reply:
Neither is catching a checked exception. Anyway, `getAndIncrement` is already using CAS, isn't?

Reply:
OK, but isn't `getAndIncrement` already using CAS ?

Reply:
```
        int p = pos.getAndIncrement();
        if (p >= len) {
          p = p % len;
          if (p == 0) {
            pos.addAndGet(-len);
          }
        }
        return (T) elements[p];
```
is more accurate (correct distribution) and scale better concurrently across different sizes of holders (tested from 2 to 10)

Reply:
ok